### ISSUE TITLE
Upgraded from 0.7.0 to 2.1.0, static resource serving needs full restart

### ISSUE THREAD

**Issue by alexflav23** [earldouglas/xsbt-web-plugin#298]
After many years of being loyal to `0.7.0` as a result of copy paste setting up builds, I've finally decided to upgrade to `2.1.0` so I can properly specify JVM options for the container.

Much to my horror, simple HTML pages with new content now require a full server restart to happen as any other static file update does. If I lacked spare time before, this would eliminate the little I have left, so I'm hoping I've somehow messed up the configuration really bad.

Please confirm how this can be resolved, I'd expect simple browser refreshes to serve the latest version of a static resource.

**Comment by earldouglas** [comment]
The problem with serving source code directly is it would circumvent post-processors such as [`webappPostProcess`](https://github.com/earldouglas/xsbt-web-plugin/blob/master/docs/2.1.md#massaging-the-war-file), so you can imagine it would be complex to specify configuration that accommodates both approaches.

How would you propose configuring xsbt-web-plugin to distinguish between static and non-static resources?

**Comment by alexflav23** [comment]
Hi @earldouglas,

I'd define a consensus based on path and file extension, kind of like the Play framework does. Everything static should go under `src/main/assets/` lets say. Of course you can have a sensible default and allow this to be configurable.

And you compile assets to `/src/main/webapp/assets`, so everything under there is safely deemed fully static. I've seen this used a number of times, perhaps it would be a sensible default. You can either add an in-between phase that gets triggered for compilation of assets, or users can manually `compile` and then look at the latest compiled JS files for instance with a single browser refresh, just like the old days of `0.7.0`.

I appreciate it's not the kind of issue that can be resolved trivially, but I think delay wise I'd expect close to instant updating, a full container restart with the load of stuff that a big app runs on Boot is definitely a lot to look at between adding two paragraphs or single CSS changes.

Regards.

**Comment by earldouglas** [comment]
These are interesting ideas; thanks for the suggestions.  I know it gets frustrating when development cycle times are long, so any ways to optimize the plugin are worth consideration.

The tricky part is balancing optimization with things like arbitrary *.war* file massaging and maintaining a production-like development environment (i.e. deploying a *.war* file, rather than source code).

I'll see if there's a way we can't accommodate both, and let the developer choose which they prefer.

**Comment by alexflav23** [comment]
HI @earldouglas,

Has there been any progress on this or anything the community can do to help? I will admit time is not exactly a luxury on my hands so I would hate to over-commit there, but it would be interesting to have the steps in implementing this duo-mode behaviour outlined somewhere.

I would dare to believe a couple of people using this plugin actively are incentivised to see the the feature implemented, perhaps with you chairing the process it could be done.

Regards.

**Comment by atais** [comment]
+1

**Comment by earldouglas** [comment]
@alexflav23 Awesome idea.  How can I help?

**Comment by atais** [comment]
@earldouglas I think it would be great if it would be possible to mark static resources, possibly the same way `watchSources` are managed at the moment? 

```
watchSources := watchSources.value.filterNot { x =>
      x.isDirectory || x.getAbsolutePath.contains("webapp")
    }
```
or you could do all type of extension matching here.

Later, I believe, the static resources changes should be updated on tomcat but without the restart of the whole container. 

What do you think?

**Comment by alexflav23** [comment]
Hi @earldouglas,

Assuming a third party knows nothing about the plugin internals, it would be nice to know:

- What are the relevant bits of code to look at with respect to the pipeline of packaging.
- What are the things that we need to do to enable static resource access.

**Comment by atais** [comment]
After quite long time I finally had a moment to look into the issue. I would like to resolve it completely, but it will be difficult without your assist @earldouglas.

----------------
I have prepared a simple modification to the plugin, that basically serves all the content from `/src/main/webapp/` instead of `/target/webapp/`. Using it with configuration like:
```scala
watchSources := watchSources.value.filterNot { x =>
   x.isDirectory || x.getAbsolutePath.contains("webapp")
}
```
resulted in direct access to static sources, and immediate refresh working with no recompilation.

The current version is available at my fork: https://github.com/atais/xsbt-web-plugin
After downloading you can install it locally with `sbt publishLocal` and your other projects can use it as `2.2.0` version.

**However** I understand my situation may be different from others. We use `gulp` for frontend build operations and all those actions take place in `/src/main/webapps` and save results in the same folder. **What is the situation with `sbt-less` or `sbt-js`**?

--------------
During the modifications I have realized how complex the process is and why it is difficult to fix.

1. (Currently) Tomcat serves all the content available in `/target/webapp/` which is **copied** there during build time. Though development is made in `/src/main/webapp/`.
2. The sources (f.e. `/src/main/webapp/`) are watched with build-in `sbt` `watchSources`. Any change results in re-build (and also - re-copy the static files).
3. Ignoring those files in `watchSources` will cause them not to be copied when they change.
4. I am not sure if serving content from `/src/main/webapp/` is a good idea, but (for me) it works.

Knowing the above, @earldouglas:
1. How were the files served in `0.7.0` that the reload did work?
2. Is there other mechanism to watch static files we could use to copy them on change, but not affecting `sbt` `watchSources`? 
3. ??? Other ideas are welcome.

I could make a PR for this but I do not think current solution is ready for the release.

**Comment by earldouglas** [comment]
@atais This looks like a pretty clean approach to the problem.  What if we made your changes selectable by a new setting, something like `developmentMode := true`?

One thing to watch out for is that we currently allow [arbitrary manipulation of the `.war` file](https://github.com/earldouglas/xsbt-web-plugin/blob/master/docs/2.1.md#massaging-the-war-file) at build time, so with a "development mode" it would be impossible to support this in the same way.

This might be fine, as long as users understand that they can't use both at the same time.

> How were the files served in 0.7.0 that the reload did work?

The biggest difference is that xsbt-web-plugin used to launch an embedded Jetty/Tomcat server directly from its own code, in the same JVM that sbt itself runs.  There were sufficient drawbacks to this approach that we decided to fork the JVM and launch the container completely separately.  To make this as "production-like" as possible, we first build the `.war` file as we would for distribution, then pass it along to a new Jetty/Tomcat process in a forked JVM.

**Comment by atais** [comment]
@earldouglas I'm glad you like it :smile: 

Time is not in our favor (nearly weekend for me), but I will try to prepare a PR.
Could you please tell me what kind of code formatter you are using, because mine changes the whole file (build-in Intellj).

**Comment by earldouglas** [comment]
Fixed by #308.
